django/db/models/functions/text.py

```diff
@@ -73,6 +73,23 @@ class ConcatPair(Func):
 
     function = "CONCAT"
 
+    def _resolve_output_field(self):
+        text_field = None
+        max_length = 0
+        for field in self.get_source_fields():
+            if isinstance(field, TextField):
+                text_field = field
+            elif (
+                isinstance(field, CharField)
+                and (field_max_length := field.max_length) is not None
+            ):
+                max_length += field_max_length
+            else:
+                return super()._resolve_output_field()
```

Comment:
If we're dealing with anything but a `TextField` and a `CharField` with an explicit `max_length` refuse the temptation to guess.